src/grad.jl

```diff
@@ -35,15 +35,66 @@ end
 jacobian(fdm, f, x::Vector{<:Real}) = jacobian(fdm, f, x, length(f(x)))
 
 """
-    jvp(fdm, f, x::AbstractVector{<:Real}, ẋ::AbstractVector{<:Real})
+    _jvp(fdm, f, x::Vector{<:Real}, ẋ::AbstractVector{<:Real})
 
 Convenience function to compute `jacobian(f, x) * ẋ`.
 """
-jvp(fdm, f, x::Vector{<:Real}, ẋ::AV{<:Real}) = jacobian(fdm, f, x) * ẋ
+_jvp(fdm, f, x::Vector{<:Real}, ẋ::AV{<:Real}) = jacobian(fdm, f, x) * ẋ
 
 """
-    j′vp(fdm, f, ȳ::AbstractVector{<:Real}, x::AbstractVector{<:Real})
+    _j′vp(fdm, f, ȳ::AbstractVector{<:Real}, x::Vector{<:Real})
 
 Convenience function to compute `jacobian(f, x)' * ȳ`.
 """
-j′vp(fdm, f, ȳ::AV{<:Real}, x::Vector{<:Real}) = jacobian(fdm, f, x, length(ȳ))' * ȳ
+_j′vp(fdm, f, ȳ::AV{<:Real}, x::Vector{<:Real}) = jacobian(fdm, f, x, length(ȳ))' * ȳ
+
+"""
+    jvp(fdm, f, x, ẋ)
+
+Compute a Jacobian-vector product with any types of arguments for which `to_vec` is defined.
+"""
+function jvp(fdm, f, (x, ẋ)::Tuple{Any, Any})
+    x_vec, vec_to_x = to_vec(x)
+    _, vec_to_y = to_vec(f(x))
+    return vec_to_y(_jvp(fdm, x_vec->to_vec(f(vec_to_x(x_vec)))[1], x_vec, to_vec(ẋ)[1]))
+end
+function jvp(fdm, f, xẋs::Tuple{Any, Any}...)
+    x, ẋ = collect(zip(xẋs...))
```

Comment:
I'm having a difficult time following this part of the code. Note that in the function signature, you're restricting `xẋs` to be a 2-tuple and splatting it, which means that the signature is just
```julia
function jvp(fdm, f, x, ẋ)
```
Now, the line below the signature is equivalent to
```julia
x, ẋ = collect(zip(x, ẋ))
```
If `x` or `ẋ` has more than 2 elements, this is discarding the rest, which means this is equivalent to
```julia
x, ẋ = (x[1], ẋ[1]), (x[2], ẋ[2])
```
Then so what's getting passed to the next call is a tuple of 2-tuples.

Is that analysis correct and the behavior is intentional, or have I missed something here?

Reply:
Hmm possibly I've messed up here. This signature is meant to be shorthand for
```julia
function jvp(fdm, f, (x, xdot), (y, ydot), (z, zdot), ...)
```
so that each `Tuple{Any, Any}` contains an argument-differential pair. As a result, the (un)zipping operation should produce
```julia
x, ẋ = (x, y, z), (xdot, ydot, zdot)
```

Reply:
Ah, scratch that, got the wrong thing to mind. Your analysis is correct. Not that I've also wrapped `f` in an anonymous function that splats out the arguments. This makes it possible to handle varargs functions by converting them into unary functions of `Tuple`s.